tests/test_finite_difference.py

```diff
@@ -7,43 +7,37 @@ def setup_method():
     np.random.seed(0)
 
 
-@pytest.mark.parametrize(
-        'x_data', [np.arange(10), 3]
-)
-def test_model(x_data):
-    x, y, z = sf.variables('x, y, z')
-    a, b, c = sf.parameters('a, b, c')
-
-    '''Tests the case with 1 component and 1 parameter'''
-    model = sf.Model({y: 3 * a * x**2})
-    exact = model.eval_jacobian(x=x_data, a=3.5)
-    approx = model.finite_difference(x=x_data, a=3.5)
-    _assert_equal(exact, approx)
-
-    '''Tests the case with 1 component and multiple parameters'''
-    model = sf.Model({y: 3 * a * x**2 - sf.exp(b) * x})
-    exact = model.eval_jacobian(x=x_data, a=3.5, b=2)
-    approx = model.finite_difference(x=x_data, a=3.5, b=2)
-    _assert_equal(exact, approx)
+x, y, z = sf.variables('x, y, z')
+a, b, c = sf.parameters('a, b, c')
```

Comment:
Can you try moving these to conftest.py (?) There is a "magic" pytest file that gets run during test setup. Alternatively, create a separate file with variables/parameters and import that.
This is fine here because it's just one file, but we'll need these in every single test...

Reply:
The name of the "magic" file is `conftest.py` :D Now we have some options:

1. We could add the variables and parameters to the pytest namespace, like:
```python
import pytest
import symfit as sf

def pytest_configure():
    pytest.x = sf.Variable('x')
    pytest.y = sf.Variable('y')
```
as `conftest.py` and the test example
```python
import pytest
import symfit as sf

def test_example():
   model = sf.Model({pytest.y: pytest.x})
   assert someting
```

2. We could use another file for the variables and parameters:
```python
import symfit as sf
x, y = sf.variables('x, y')
a, b, c = sf.parameters('a, b, c')
```
and import this file to use it
```python
from var import * # import the special file
import symfit as sf

def test_example():
   model = sf.Model({y: a*x**b+c})
   assert someting
```
Personally, I would prefer the second option, unless you know how to get `pytest.something` away :)

Reply:
As it stands I also prefer the second option.
However, if there's a way to make pytest re-initialize the variables/parameters for every test that wins. But I don't know if that can be done (and a quick search doesn't reveal anything).

Reply:
The probability is high that there is an option for fixtures to be called each time. You could also introduce global fixtures that are accessible to each file. However, every variable/parameter (with a bit of magic even just one variable) would have to be passed as a function argument. 

Either way, an additional file is the most elegant way.

Reply:
The probability is high that there is an option for fixtures to be called each time. You could also introduce global fixtures that are accessible to each file. However, every variable/parameter (with a bit of magic even just one variable) would have to be passed as a function argument + We still can't add fixtures to parametrizes. 

Either way, an additional file is most elegant. During the first review, I noticed that some min/max values are set. However, in many cases, these have no influence, for example, https://github.com/tBuLi/symfit/blob/bc25c8c70d176f0cbcc713143e160bb0cb96778f/tests/test_finite_difference.py#L67-L68 
Can I remove boundaries, if they aren't necessary?

Reply:
> But we can't use them in every test, because we're setting max/min/default values...

Which would be why I prefer something that magically reinitializes them for every test. I guess we'll just have to set a min/max/value in every test, maybe with a `reset_vars` function.

Reply:
That's the idea of fixtures, but we can't use them in parametrizes. :/
I'll do some research before I implement the collections.

Reply:
We can use fixtures not to declare one variable but to re-initialize every global variable and parameter. And with `@pytest.fixture(autouse=True)` the fixture is automatically included in every test. 
The ___special file___ would be something like:
```python
import symfit as sf
import pytest

w, x, y, z = sf.variables('w, x, y, z')
a, b, c = sf.parameters('a, b, c')


@pytest.fixture(autouse=True)
def initVariales():
    global w, x, y, z
    w, x, y, z = sf.variables('w, x, y, z')


@pytest.fixture(autouse=True)
def initParameters():
    global a, b, c
    a, b, c = sf.parameters('a, b, c')
```

Reply:
This looks like what we want :)

> And with @pytest.fixture(autouse=True) the fixture is automatically included in every test.

This includes every single parametrization, right?

Great find 👍

Reply:
> This includes every single parametrization, right?

Yes, it does. I tested it :D It reinitializes the variables and parameters before every parameterization
> Great find 👍

Thanks :))

Reply:
Maybe we should discuss about the special file with the collection... Because I have the impression that such a collection is more irritating than helpful.
My pros and cons:

**Pros**
- shorter tests
- Use parameters and variables in parametrizations

**Cons**
- difficult/not possible to convert complex tests with complex physical equation system without losing track
- _maybe_ difficult for new devs to write new tests
- Losing overview over variables and parameters in a complex test without the declaration

Perhaps a large collection of variables is not the ideal solution.  But for the situation where you want to use variables and parameters in a parametrization, we can introduce an adjusted pattern, such as:
```python
import pytest
import symfit as sf

a, b, c = sf.parameters('a, b, c')
x, y, z = sf.variables('x, y, z')

@pytest.fixtrue
def initVariables():
    global x, y, z
    x, y, z = sf.variables('x, y, z')


@pytest.fixtrue
def initParameters():
    global a, b, c
    a, b, c = sf.parameters('a, b, c')


@pytest.mark.parametrize('model', 
    [
        {z: a * x + b * y + c},
        {z: x**a + y**b + c}
    ]
)
def test_something(initVariables, initParameters, model):
    # do something with the models
```

Reply:
I think I still prefer the autofixture solution: it will only affect the global abcxyz, and not any that you create in specific tests.

> difficult/not possible to convert complex tests with complex physical equation system without losing track

Why? With these cases you can always instantiate the parameters and variable in the test itself. But it should be a hint the test is too complex in the first place.

> maybe difficult for new devs to write new tests

Easier than the alternative, where parameters don't get reset and your tests become history dependant. In addition, this mechanism is opt-in: if you don't have parameters in the global namespace it does not affect you

> Losing overview over variables and parameters in a complex test without the declaration
I'd argue the opposite: the declaration is not there to distract you from the actual test.

Reply:
> But it should be a hint the test is too complex in the first place.

I mean "complex tests" in the sense of using real examples from physics, where Greek terms are used.

> Easier than the alternative, where parameters don't get reset and your tests become history dependant

In case of the described pattern or IN tests with parameterization, the parameters are reset btw :D (autouse allows you to dispense with entering the variable as a function argument)

> I'd argue the opposite: the declaration is not there to distract you from the actual test.

In this case, the tests aren't like a book, that you can simply read...you have to look up, how *a* or *x* are declared. 

That said to my previous picture of the parameter and variable collection. The collection with a handful of variables and parameters is cleaner. But then I would only convert those tests that only access x, y, z, a, b, c. But as soon as another parameter (different from the ones listed) appears, I'd do without it.

Reply:
> I mean "complex tests" in the sense of using real examples from physics, where Greek terms are used.

But the test is about code, and not about physics, and maybe those tests can be simplified by taking out the physics to test a simpler model that exercises the same code path.

> In case of the described pattern or IN tests with parameterization, the parameters are reset btw :D (autouse allows you to dispense with entering the variable as a function argument)

```python
def test_something():
    a, b = variables('a, b')
    a.max=10
    ...
    fit(...)
    assert ...
```
The autouse fixture will never touch these a and b since they're local variables, and the autofixtures touch the *global* a, b, c.

> In this case, the tests aren't like a book, that you can simply read...you have to look up, how a or x are declared.

If the declaration matters to the test you should've declared them locally in the first place.

Start by converting only a few tests to see what this would look like (i.e. finite_difference and fit_result). And regarding parameters/variables beyond abc/xyz, I think a *lot* of tests could use simpler, non-physical, models to test the same behaviour.

Reply:
Then I'll start with those :D
And converting complex physical tests into simpler ones could be a new issue...